vercors/src/main/java/vct/col/rewrite/RewriteArrayRef.java

```diff
@@ -3,269 +3,391 @@
+    return res;
   }
   
   @Override
-	public void visit(OperatorExpression e){
-		switch (e.operator()){
-    case ValidArray:{
-      ASTNode M=rewrite(e.arg(0));
-      ASTNode MD=create.expression(StandardOperator.OptionGet,M);
-      ASTNode sz=rewrite(e.arg(1));
-      result=create.expression(StandardOperator.And
-        ,create.expression(StandardOperator.NEQ,M,create.reserved_name(ASTReserved.OptionNone))
-        ,and(create.expression(StandardOperator.EQ,create.expression(StandardOperator.Size,MD),sz),
-             injective(MD)
-            )
-      );
+  public void visit(OperatorExpression e) {
+    switch (e.operator()) {
+    case ValidArray: {
+      ASTNode M = rewrite(e.arg(0));
+      ASTNode sz = rewrite(e.arg(1));
+      // Determine type in order to generate appropiate conditions for ValidArray.
+      Type t = e.first().getType();
+      t = rewrite(t);
+      M.setType(t);
+      result = validArrayFor(M, sz);
       break;
     }
-    case ValidMatrix:{
-      ASTNode M=rewrite(e.arg(0));
-      ASTNode MD=create.expression(StandardOperator.OptionGet,M);
-      ASTNode sz=create.expression(StandardOperator.Mult,
-          rewrite(e.arg(1)),rewrite(e.arg(2)));
-      result=create.expression(StandardOperator.And
-        ,create.expression(StandardOperator.NEQ,M,create.reserved_name(ASTReserved.OptionNone))
-        ,and(create.expression(StandardOperator.EQ,create.expression(StandardOperator.Size,MD),sz),
-             injective(MD)
-            )
-      );
+    case ValidMatrix: {
+      ASTNode M = rewrite(e.arg(0));
+      ASTNode sz1 = rewrite(e.arg(1));
+      ASTNode sz2 = rewrite(e.arg(2));
+      // Determine type in order to generate appropiate conditions for ValidMatrix.
+      Type t = e.first().getType();
+      t = rewrite(t);
+      M.setType(t);
+      result = validMatrixFor(M, sz1, sz2);
       break;
     }
-		  case EQ:
-		  case NEQ:
-		  {
-        ASTNode e0=e.arg(0);
-        ASTNode e1=e.arg(1);
-        ASTNode array=null;
-        if (e0.isReserved(ASTReserved.Null) && e1.getType().isPrimitive(PrimitiveSort.Array)){
-          array=e1;
-        }
-        if (e1.isReserved(ASTReserved.Null) && e0.getType().isPrimitive(PrimitiveSort.Array)){
-          array=e0;
+    case EQ:
+    case NEQ: {
+      ASTNode e0 = e.arg(0);
+      ASTNode e1 = e.arg(1);
+      ASTNode array = null;
+      if (e0.isReserved(ASTReserved.Null) && e1.getType().isPrimitive(PrimitiveSort.Array)) {
+        array = e1;
+      }
+      if (e1.isReserved(ASTReserved.Null) && e0.getType().isPrimitive(PrimitiveSort.Array)) {
+        array = e0;
+      }
+      if (array != null && rewrite(array.getType()).isPrimitive(PrimitiveSort.Option)) {
+        result = create.expression(e.operator(), create.reserved_name(ASTReserved.OptionNone), rewrite(array));
+      } else {
+        super.visit(e);
+      }
+      break;
+    }
+    case Length:
+      if (e.first().getType().isPrimitive(PrimitiveSort.Array)) {
```

Comment:
This should probably be the other way around, no? If it is an option you should generate a get, when it is an array you shoud not.